### DISCUSSION TITLE
Error upon 'make' command

### DISCUSSION THREAD

**Discussion by Pkaps25** [ChaissonLab/danbing-tk#1] [Q&A] [answered]
Hi, I just cloned the repo on a linux machine. Upon executing the setup instructions, I get this error with the command `make` inside of the danbing-tk directory:

`g++ -std=c++11 -lpthread -O3 -o bin/danbing-tk src/aQueryFasta_thread.cpp
/bin/ld: cannot open output file bin/danbing-tk: No such file or directory
collect2: error: ld returned 1 exit status
make: *** [bin/danbing-tk] Error 1
`

**Comment by joyeuxnoel8** [comment, marked as answer]
Hi, 
Thanks for reporting the error. I've updated the Makefile. Let me know it there's still any error message.
Tsung-Yu

**Reply by Pkaps25** [reply]
Thank you, this fixed the error

**Comment by Pkaps25** [comment]
I'm experiencing another error when following the test instructions. I've replaced PREFIX in the bam.fofn file and in the config file, but it appears a file is missing from the test inputs. Could you clarify what the format of this input file should look like please.

`OSError in line 11 of /local/workdir/pk445/vntr/danbing-tk/pipeline/GoodPanGenomeGraph.snakefile:
/workdir/pk445/vntr/danbing-tk/test/input/genome.bam.tsv.bak not found.
  File "/local/workdir/pk445/vntr/danbing-tk/pipeline/GoodPanGenomeGraph.snakefile", line 11, in <module>
  File "/usr/lib64/python3.6/site-packages/numpy/lib/npyio.py", line 961, in loadtxt
  File "/usr/lib64/python3.6/site-packages/numpy/lib/_datasource.py", line 195, in open
  File "/usr/lib64/python3.6/site-packages/numpy/lib/_datasource.py", line 535, in open
`

**Comment by joyeuxnoel8** [comment]
Removed obsolete files and added the missing file. Now you just need to replace $PREFIX in `goodPanGenomeGraph.json` and `input/genome.bam.tsv`. Let me know if it works. Thanks!

**Reply by Pkaps25** [reply]
This fixed the error. I had to keep the `genomes.txt` file to avoid a different error, however.

**Reply by joyeuxnoel8** [reply]
Glad to know you were able to get it running! I now removed the dependency for `genomes.txt`.

**Comment by joyeuxnoel8** [comment]
Hello,

I can't seem to find your comment on the page. I guess you've got it
figured out; I'll make sure to fix that as well. Also, I'd like to let you
know that I'm working on a release that will update the RPGG and the
annotated VNTR set since it looks like you're trying to genotype samples.

**Reply by Pkaps25** [reply]
Thank you for letting me know and for your help. I've been working on an analysis using my data and an intermediate file is missing information. I've attached the error trace and snippets of temp file with missing data and another temp file that has this data. Files named tmp1.0.bed and tmp1.1.bed are missing the data, but files named tmp0.0.bed and tmp0.1.bed have the required columns. Do you have any suggestions on what I can do?

The error from snakemake:
[2021-01-11T172422.186716.snakemake.log](https://github.com/ChaissonLab/danbing-tk/files/5798612/2021-01-11T172422.186716.snakemake.log)


A snippet of the temp bed file with data present:
[good.txt](https://github.com/ChaissonLab/danbing-tk/files/5798578/good.txt)

A snippet of the temp bed file that is causing the error:		
[bad.txt](https://github.com/ChaissonLab/danbing-tk/files/5798581/bad.txt)

**Reply by joyeuxnoel8** [reply]
This is very likely due to the `_` character in contig names such as `NZ_CP011510.1`. And just double check, are you using your own reference (not hg38) and VNTR annotations? 
The `tmp0.l0.bed` and `tmp0.l1.bed` files are generated by liftover and should have the format ```asm_name    asm_start   asm_end  ref_name    ref_start ref_end    strand```. 
And from the `good.txt`, or  `tmp0.0.bed`,  you provided, something was off during the post-processing step in `script/liftbed.clean.py`. I need to generalize the code to non-hg38 reference if that's the case.

**Reply by Pkaps25** [reply]
I am working with tuberculosis data, so I am using the H37Rv tuberculosis genome and using tuberculosis VNTR annotations. It appears the `_` character is causing problems in the `tmp0.l0.bed` and `tmp0.l1.bed` files. I will remove the `_` character and rename the fields and let you know. Will the pipeline currently support using the TB reference genome?

**Reply by joyeuxnoel8** [reply]
I tend to avoid using `_` in fasta header in my pipeline, but this might not be very user-friendly. In principle, any genomes and VNTRs should work, except for some edge cases to consider such as the fasta header. One thing to note is that the current pipeline has a quite stringent quality control on VNTR size, number of supporting haplotypes, etc. You might need to adjust the config file parameters accordingly if you really need every VNTR to be included in the RPGG. 

And yes, please do let me know how it goes. That would be very helpful for other users!

**Reply by Pkaps25** [reply]
I was able to run the pipeline by removing the `_` character. The file `pan.tr.kmers` contained the numbers of the loci without any associated information, which leads me to believe that the default QC will not work for my dataset. Thanks for your help!

**Reply by joyeuxnoel8** [reply]
Hmm... all loci in `pan.tr.mbe.v2.bed` and `pan.tr.kmers` should already have decent quality. The empty loci (no kmer information) should be due to some other reason. I'd be happy to help look into it if you can provide more information, e.g. `pan.tr.mbe.v2.bed`, `goodPanGenomeGraph.json`.

**Reply by Pkaps25** [reply]
Here are the two files (Converted to `.txt` to upload to github):
 `pan.tr.mbe.v2.bed` 
[pan.tr.mbe.v2.txt](https://github.com/ChaissonLab/danbing-tk/files/5804550/pan.tr.mbe.v2.txt)

`goodPanGenomeGraph.json`
[goodPanGenomeGraph.txt](https://github.com/ChaissonLab/danbing-tk/files/5804553/goodPanGenomeGraph.txt)

Please let me know if this helps. The resulting `pan.tr.kmers` file looks like this:

\>0
\>1
\>2
...
\>11

**Reply by joyeuxnoel8** [reply]
Yes, this helps. The VNTR annotations did succeed. So the problem might be during the `script/SelectRegions.py` step since I forgot to include `pysam` as required package in conda installation. Is the `*.tr.fasta` empty for all genomes?

**Reply by Pkaps25** [reply]
The `*.tr.fasta` files for all of the genomes are not empty.

**Reply by joyeuxnoel8** [reply]
Great, so `* .rawPB.tr.kmers` should be non-empty as well then. What's the approximate coverage of your short-read sequencing data? Could you also provide one of your `* .rawPB.tr.kmers` and `* .rawIL.tr.kmers`?

**Reply by Pkaps25** [reply]
I believe I am not inputting the correct data to the pipeline. As of right now I am using assembled TB genomes in place of LRS and in place of SRS data. The bam files I am using are the result of aligning the TB genomes to the TB reference. I think this is causing the problem.

**Reply by joyeuxnoel8** [reply]
Ah, good catch! Just one note, LRS data is not required; only the assembled genomes are. The bam files are to ensure the quality of the assemblies are good enough for downstream analysis. I might need to make this optional to accommodate different uses of the RPGG.

**Reply by Pkaps25** [reply]
I want to sanity check my approach. Please let me know if I am thinking about this correctly. The only data I need to make the RPGG are the assembled genomes and the reference genome? For the assembled genomes, I plan on using complete RefSeq genomes for 3 TB strains (for a test run) and adding more genomes later. In this case, can my BAM files contain the result of aligning the complete genomes to the reference? I notice that the pipeline expects two fasta files for each genome `*.0.fa` and `*.1.fa`. If I am using the RefSeq Genomes, is it acceptable to use the same fasta file for both `*.0.fa` and `*.1.fa`? After creating the RPGG, I plan to align cfDNA reads to the graph.

**Reply by joyeuxnoel8** [reply]
> The only data I need to make the RPGG are the assembled genomes and the reference genome?

I probably didn't document this well, but the BAM files here are from SRS data matched with each genome and not from SRS or assemblies. Assembly bam files are used for liftover and are generated by the pipeline by aligning the specified reference to each assembly. I've updated the danbing-tk build section in [README](https://github.com/ChaissonLab/danbing-tk/blob/master/README.md#danbing-tk-build) for clarity. If you do not have whole-genome SRS data, let me know. I will make the graph pruning step optional.

> I notice that the pipeline expects two fasta files for each genome *.0.fa and *.1.fa. If I am using the RefSeq Genomes, is it acceptable to use the same fasta file for both *.0.fa and *.1.fa?

For testing purpose, this should be fine. And for un-phased assemblies, this is also the easy way to trick the program.

> After creating the RPGG, I plan to align cfDNA reads to the graph.

This is cool! I look forward to the application.

**Reply by Pkaps25** [reply]
Thank you. I don't have the whole genome SRS data. If you could make the pruning optional that would be great. 

For posterity, I tried simulating illumina short reads (40x and 80x) from each of the genomes and aligning those simulated reads back to their respective genomes to use as bam inputs. The output `pan.tr.kmers` file was still empty.

**Reply by joyeuxnoel8** [reply]
Hmm this is weird. Could you show an example of the paired-end reads?

**Reply by Pkaps25** [reply]
Apologies, please disregard that last comment. The above approach produces an output in the `pan.tr.kmers` file.

**Reply by joyeuxnoel8** [reply]
Glad to know this! Also the SRS data, or the pruning step, is optional now. Please check [README](https://github.com/ChaissonLab/danbing-tk/blob/master/README.md#danbing-tk-build) for updated usage.
